pkg/oci/build.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/docker/distribution/reference"
+	"github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+)
+
+const (
+	ARCH_AMD64 = "amd64"
+	ARCH_ARM64 = "arm64"
+)
+
+type BuildGadgetImageOpts struct {
+	// List of eBPF objects to include on the image. The key is the architecture and the value
+	// is the path to the eBPF object.
+	EBPFObjectPaths map[string]string
+	// Path to the medatada file
+	MetadataPath string
+}
+
+// BuildGadgetImage creates an OCI image with the objects provided in opts. The created image is
+// tagged if image is not empty.
+func BuildGadgetImage(opts *BuildGadgetImageOpts, image string) (*GadgetImageDesc, error) {
+	ociStore, err := getLocalOciStore()
+	if err != nil {
+		return nil, fmt.Errorf("getting oci store: %w", err)
+	}
+
+	indexDesc, err := createImageIndex(ociStore, opts)
+	if err != nil {
+		return nil, fmt.Errorf("creating image index: %w", err)
+	}
+
+	imageDesc := &GadgetImageDesc{
+		Digest: indexDesc.Digest.String(),
+	}
+
+	if image != "" {
+		targetImage, err := normalizeImageName(image)
+		if err != nil {
+			return nil, fmt.Errorf("normalizing image: %w", err)
+		}
+
+		err = ociStore.Tag(context.TODO(), indexDesc, targetImage.String())
+		if err != nil {
+			return nil, fmt.Errorf("tagging manifest: %w", err)
+		}
+
+		imageDesc.Repository = targetImage.Name()
+		if ref, ok := targetImage.(reference.Tagged); ok {
+			imageDesc.Tag = ref.Tag()
+		}
+	}
+
+	return imageDesc, nil
+}
+
+func pushDescriptorIfNotExists(target oras.Target, desc ocispec.Descriptor, contentReader io.Reader) error {
+	err := target.Push(context.TODO(), desc, contentReader)
+	if err != nil && !errors.Is(err, errdef.ErrAlreadyExists) {
+		return fmt.Errorf("pushing descriptor: %w", err)
+	}
+	return nil
+}
+
+func createEbpfProgramDesc(target oras.Target, progFilePath string) (ocispec.Descriptor, error) {
+	progBytes, err := os.ReadFile(progFilePath)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("reading eBPF program file: %w", err)
+	}
+	progDesc := content.NewDescriptorFromBytes("application/vnd.gadget.ebpf.program.v1+binary", progBytes)
```

Comment:
Maybe use consts for `application/vnd.gadget.ebpf.program.v1+binary` and `application/vnd.gadget.config.v1+yaml` as well?